### PR TITLE
py-kornia: add v0.7.2

### PR DESCRIPTION
**Pull request by adamjstewart**
https://github.com/kornia/kornia/releases/tag/v0.7.2

### CODE REVIEW

**Comment by adamjstewart** [comment]
At the request of @johnnv1, I'm adding the following Kornia maintainers to the package recipe:

@edgarriba
@ducha-aiki
@lferraz
@shijianjian
@cjpurackal
@johnnv1

I'm also adding myself since I heavily use Kornia in TorchGeo. For those of you who have never heard of Spack, it is a package manager similar to Conda but targeted towards HPC and not limited to just Python/R. The previous spackbot comment explains the basic roles of a maintainer. I don't expect the commitment to be very high (basically just adding a new version once in a while or solving the occasional build issue). If any of you do not want this role, or only want it for kornia and not kornia-rs, let me know and I'll remove you.

**Comment by ducha-aiki** [comment]
Thank you! Is there a "Getting started doc" how to add new versions?
Or the only thing we have to do, is to create a similar PR with a new version and hash?

**Comment by edgarriba** [comment]
@adamjstewart do you guys have somehow an automatic way to detect new versions (and send notifications) or we as maintainers we need to open a pr every time we release ?

**Comment by adamjstewart** [comment]
> Thank you! Is there a "Getting started doc" how to add new versions? Or the only thing we have to do, is to create a similar PR with a new version and hash?

Spack has the fun problem that we literally have **too much** documentation. The most relevant docs to you would be the [Packaging Guide](https://spack.readthedocs.io/en/latest/packaging_guide.html) and the [PythonPackage docs](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html). There's also a [Packaging Tutorial](https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html). But honestly, most of the time you just run `spack checksum py-kornia` and it scrapes PyPI and asks you which new versions you would like to add to the package. At the moment, adding or changing dependencies is still manual, but we're also working on automating that.

**Comment by adamjstewart** [comment]
> @adamjstewart do you guys have somehow an automatic way to detect new versions (and send notifications) or we as maintainers we need to open a pr every time we release ?

We do not yet have automatic detection like conda-forge has but that's also a work in progress. The biggest problem is that Spack is centralized whereas conda-forge is distributed, so if we get hundreds of new PRs per day we can't realistically review and merge all of them. We're still figuring out ways to scale up like that. In the long run, the hope is to remove most pure-Python packages from Spack and only keep packages like numpy or kornia-rs that require compilation and non-Python dependencies. The rest would then be auto-generated on the fly using pip.

File: var/spack/repos/builtin/packages/py-kornia-rs/package.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyKorniaRs(PythonPackage):
+    """Low level implementations for computer vision in Rust."""
+
+    homepage = "https://www.kornia.org/"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-kornia-rs/package.py]: Are you sure this site supports the `https` protocol?  I can only access it via `http`.

**Review by tldahlgren — COMMENTED** [review]
Confirmed the (new) version sha256s but not the homepage.

**Comment by tldahlgren** [comment]
'spack spec' succeeds; however, there are problems with the ML stacks for the new package.

**Comment by edgarriba** [comment]
@tldahlgren where can i see the specifications of those machines ? for context this is the full matrix that we use to release the python wheel of kornia-rs. I'm seeing that for some reason in the ml stack seems that it tries to compile the library which without the right dependencies might not be a good idea.

https://github.com/kornia/kornia-rs/blob/main/.github/workflows/python_release.yml

File: var/spack/repos/builtin/packages/py-kornia-rs/package.py (hunk 1 above)
Comment (edgarriba) [reply var/spack/repos/builtin/packages/py-kornia-rs/package.py]: http is the right one

**Comment by tldahlgren** [comment]
> @tldahlgren where can i see the specifications of those machines ? for context this is the full matrix that we use to release the python wheel of kornia-rs. I'm seeing that for some reason in the ml stack seems that it tries to compile the library which without the right dependencies might not be a good idea.
> 
> https://github.com/kornia/kornia-rs/blob/main/.github/workflows/python_release.yml

The runners are generally containers.  You can get the URL of the container used by each stack by looking for the `image:name` under the `spack:ci:pipeline-gen` entry in each stack's `spack.yaml` file.  

The stacks can be found in stack-specific subdirectories under https://github.com/spack/spack/tree/develop/share/spack/gitlab/cloud_pipelines.

File: var/spack/repos/builtin/packages/py-kornia-rs/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-kornia-rs/package.py]: Changed to http. I just copied the https from [here](https://github.com/kornia/kornia/blob/d6b18a5188f3ee09447a542306c86d1c71ecf2f6/pyproject.toml#L43). This should either be changed or support for https should be added.

**Comment by adamjstewart** [comment]
> I'm seeing that for some reason in the ml stack seems that it tries to compile the library which without the right dependencies might not be a good idea.

Spack builds all packages from source by default. Normally we explicitly list all dependencies, although we don't yet have a good way to model rust/cargo dependencies (too many, require specific versions, auto-installed). The source build works fine for me on macOS arm64, so I'm not sure why it's failing in the pipeline. The error looks like:
```
  error: failed to run custom build command for `rav1e v0.7.1`
  Caused by:
    process didn't exit successfully: `/private/var/folders/6n/pw9tmf6n08dfgg5td4_w9wh00000gx/T/gitlab-runner/spack-stage/spack-stage-py-kornia-rs-0.1.1-4qozlonhsn56ztbsy2ceef65zexgrkwv/spack-src/py-kornia/target/release/build/rav1e-0bfb712d55c5d71b/build-script-build` (exit status: 101)
    --- stderr
    thread 'main' panicked at /Users/gitlab-runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rav1e-0.7.1/build.rs:250:29:
    Failed to acquire build-time information: Os { code: 2, kind: NotFound, message: "No such file or directory" }
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  warning: build failed, waiting for other jobs to finish...
  💥 maturin failed
    Caused by: Failed to build a native library through cargo
```
We could _try_ adding a dependency on rav1e and build it before kornia-rs, but I've never had this kind of problem before so I'm not sure how to handle it. The fact that it fails on both macOS and Linux is surprising.

**Comment by adamjstewart** [comment]
Issue reported in https://github.com/xiph/rav1e/issues/3369

**Comment by edgarriba** [comment]
what about enabling `cargo cross` in you build pipeline ? https://github.com/cross-rs/cross
or use some base rust images

**Comment by adamjstewart** [comment]
Hmm, we aren't cross compiling here, we're building and running everything on the same native architecture. We're currently building rust 1.76.0 from source but could also try a newer version.

@alecbcs may be able to help debug this.

**Comment by alecbcs** [comment]
@adamjstewart I'd try having `py-kornia-rs` depend on `rust+dev` and see if that helps. Perhaps there's some internal development functionality rav1e is relying on?

**Comment by adamjstewart** [comment]
I think we've now resolved all of the build errors! Thanks for everyone's help!

It's unfortunate that kornia now depends on Rust and LLVM and a bunch of other packages that are notoriously difficult to compile from source, but hopefully it's slightly faster?

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Review by tldahlgren — APPROVED** [review]
(Re)Confirmed homepages and version sha256s.

**Comment by edgarriba** [comment]
@adamjstewart dependencies will change over time -- actually in the latest version llvm is not required  anymore. In the next release we are introducing gstreamer and maybe dropping turbojpeg in favour of native rust libs which is much easier to maintain from our side.

**Pull request merged**